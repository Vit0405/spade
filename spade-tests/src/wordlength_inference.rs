use crate::snapshot_inference_error;
use spade_wordlength_inference::InferMethod;

snapshot_inference_error!(
    simple_multiple_addition_aa,
    "AA",
    r#"
        fn f(a: int<4>) -> int<0> {
          a + a + a + a
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(1) }
    "#
);

snapshot_inference_error!(
    simple_multiple_addition_ia,
    "IA",
    r#"
        fn f(a: int<4>) -> int<0> {
          a + a + a + a
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(1) }
    "#
);

snapshot_inference_error!(
    simple_subtraction_aa,
    "AA",
    r#"
        fn f(a: int<4>) -> int<0> {
          a - a + 1
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(1) }
    "#
);

snapshot_inference_error!(
    simple_subtraction_ia,
    "IA",
    r#"
        fn f(a: int<4>) -> int<0> {
          a - a + 1
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(1) }
    "#
);

snapshot_inference_error!(
    multiple_generics_aa,
    "AA",
    r#"
        fn f<a, b>(a: int<a>, b: int<b>) -> int<b> {
          a + b
        }

        entity main(clk: clock, rst: bool) -> int<0> {
          let x: int<5> = f(0, 10);
          let y: int<5> = f(1, 10);
          0
        }
    "#
);

snapshot_inference_error!(
    multiple_generics_ia,
    "IA",
    r#"
        fn f<a, b>(a: int<a>, b: int<b>) -> int<b> {
          a + b
        }

        entity main(clk: clock, rst: bool) -> int<0> {
          let x: int<5> = f(0, 10);
          let y: int<5> = f(1, 10);
          0
        }
    "#
);

snapshot_inference_error!(
    multiplication_and_addition_aa,
    "AA",
    r#"
        fn f<a, b>(a: int<4>, b: int<4>) -> int<0> {
          a * b + a + b
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(0, 0) }
    "#
);

snapshot_inference_error!(
    multiplication_and_addition_ia,
    "IA",
    r#"
        fn f<a, b>(a: int<4>, b: int<4>) -> int<0> {
          a * b + a + b
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(0, 0) }
    "#
);

snapshot_inference_error!(
    ifs_aa,
    "AA",
    r#"
        fn f<a, b>(q: bool, a: int<0>, b: int<4>) -> int<0> {
            if q { a } else { a * b + b - b + 1 }
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(false, 0, 0) }
    "#
);

snapshot_inference_error!(
    ifs_ia,
    "IA",
    r#"
        fn f<a, b>(q: bool, a: int<0>, b: int<4>) -> int<0> {
            if q { a } else { a * b + b - b + 1 }
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(false, 0, 0) }
    "#
);

snapshot_inference_error!(
    inner_functions_aa,
    "AA",
    r#"
        fn g() -> int<3> { 0 }
        fn f<a, b>(a: int<2>) -> int<0> {
            g() + a
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(0) }
    "#
);

snapshot_inference_error!(
    inner_functions_ia,
    "IA",
    r#"
        fn g() -> int<3> { 0 }
        fn f<a, b>(a: int<2>) -> int<0> {
            g() + a
        }
        entity main(clk: clock, rst: bool) -> int<0> { f(0) }
    "#
);

snapshot_inference_error!(
    wordlength_inference_aa,
    "AA",
    r#"
        fn p<N>(x: int<N>) -> int<N> { x }

        fn f(a: int<5>) -> int<7> {
          1 * p(a) + 1
        }

        entity main(clk: clock, rst: bool) -> int<7> {
          f(3)
        }
    "#
);

snapshot_inference_error!(
    wordlength_inference_ia,
    "IA",
    r#"
        fn p<N>(x: int<N>) -> int<N> { x }

        fn f(a: int<5>) -> int<7> {
          1 * p(a) + 1
        }

        entity main(clk: clock, rst: bool) -> int<7> {
          f(3)
        }
    "#
);
