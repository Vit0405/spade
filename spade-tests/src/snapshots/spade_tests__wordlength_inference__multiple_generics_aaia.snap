---
source: spade-tests/src/wordlength_inference.rs
---
fn f<a, b>(a: int<a>, b: int<b>) -> int<b> {
  a + b
}

entity main(clk: clock, rst: bool) -> int<0> {
  let x: int<5> = f(0, 10);
  let y: int<5> = f(1, 10);
  0
}


error: The wordlength isn't what we infered it to - inference got 2 bits and you claimed 0 bits
  ┌─ testinput:8:3
  │
8 │   0
  │   ^ Here 2 bits were infered


